### PR TITLE
Fuzz: a more efficient descriptor parsing target

### PR DESCRIPTION
**Pull request by darosior**
The current descriptor parsing fuzz target requires valid public or private keys to be provided. This is unnecessary as we are only interested in fuzzing the descriptor parsing logic here (other targets are focused on fuzzing keys serializations). And it's pretty inefficient, especially for formats that need a checksum (`xpub`, `xprv`, WIF).

This introduces a new target that mocks the keys as an index in a list of precomputed keys. Keys are represented as 2 hex characters in the descriptor. The key type (private, public, extended, ..) is deterministically based on this one-byte value. Keys are deterministically generated at target initialization. This is much more efficient and also largely reduces the size of the seeds.
TL;DR: for instance instead of requiring the fuzzer to generate a `pk(xpub6DdBu7pBoyf7RjnUVhg8y6LFCfca2QAGJ39FcsgXM52Pg7eejUHLBJn4gNMey5dacyt4AjvKzdTQiuLfRdK8rSzyqZPJmNAcYZ9kVVEz4kj)` to parse a valid descriptor, it just needs to generate a `pk(03)`.

Note we only mock the keys themselves, not the entire descriptor key expression. As we want to fuzz the real code that parses the rest of the key expression (origin, derivation paths, ..).

This is a target i used for reviewing #17190 and #27255, and figured it was worth PR'ing on its own since the added complexity for mocking the keys is minimal and it could help prevent introducing bugs to the descriptor parsing logic much more efficiently.

### CODE REVIEW

**Comment by darosior** [comment]
In what you suggest `MockDescriptor` would basically have to reimplement all of the descriptor parsing logic to be able to detect when is a key expected and replace the hex-encoded byte by an actual key. I figured the slight modification to the descriptor code to make key parsing mockable was preferable. (Note that's what we have already in Miniscript which allows us to have an efficient `miniscript_string` fuzz target.)

**Comment by dergoegge** [comment]
Concept ACK

Mocking seems fine to me, but I wonder if we could achieve something similar by placing (in)valid encoded keys in a fuzz dictionary (e.g. https://github.com/bitcoin-core/qa-assets/pull/122), which is also almost the same as Marco's suggestion. This would be less efficient compared to what is in this PR but the inputs would still be available for debugging over RPC. We can of course also just do both since adding to the dictionaries is very easy.

**Comment by darosior** [comment]
Thanks both for throwing in ideas. Note i don't have strong opinions here either, it's just some review code that i figured could be helpful having in too. However i still don't think the approaches suggested here would be better:
- Inserting valid keys at random positions in the input. I'm assuming you describe something that looks like 1) get the number of keys from the fuzzer output 2) get the positions to insert each at from the fuzzer output. I may be underestimating the fuzzer's capabilities but it sounds much less efficient.
- Including valid keys in a dictionary. Again, i'm not very familiar with the inernals of fuzzing engines but i suspect this would make the fuzzer grind the key with virtually no chance of finding another valid one, wasting a lot of cycles.
- I don't think keeping the raw seed human readable for debugging over RPC should generally be a goal, especially not at the expense of efficiency. To get something readable you can simply run `./src/test/fuzz/fuzz ./crash-XXXXX` with a `ToString()` printed to stdout.

**Comment by darosior** [comment]
Hmm good point. Using a distinguishable character fixes the issue of having to "basically reimplement descriptor parsing logic".
------- Original Message -------

**Comment by darosior** [comment]
I don't, but i think Pieter settled the debate anyways. :)
------- Original Message -------

**Comment by darosior** [comment]
Updated following @sipa's suggestion of using a marker to be able to search and replace in the descriptor string directly.

The diff is now twice smaller, and this does not touch the descriptor parsing logic anymore. After running the new `mocked_descriptor_parse` target for half a hour on my laptop i get more lines, functions and branch coverage for `descriptor.cpp` than with the existing corpus for `descriptor_parse`:
-[ `descriptor_parse`](http://download.darosior.ninja/dump/desc_coverage_not_mocked/src/script/index.html): 
image
- [`mocked_descriptor_parse`](http://download.darosior.ninja/dump/desc_coverage_mocked/src/script/index.html): 
image

File: src/test/fuzz/descriptor_parse.cpp [hunk 1]
```diff
@@ -3,11 +3,109 @@
...
+        uint8_t key_data[32] = {1};
+        // Generate keys of all kinds and store them in the keys array.
+        for (uint8_t i{0}; i < TOTAL_KEYS_GENERATED; i++) {
+            key_data[31] = i;
+
+            // If this is a "raw" key, generate a normal privkey. Otherwise generate
+            // an extended one.
+            if (IdIsCPubKey(i) || IdIsXOnlyPubKey(i) || IdIsConstPrivKey(i)) {
+                CKey privkey;
+                privkey.Set(key_data, key_data + 32, true);
```
Comment (achow101) [inline src/test/fuzz/descriptor_parse.cpp:47]: In 845810d94df83747634a60ff59635a670bba1124 "fuzz: add a new, more efficient, descriptor parsing target"

Perhaps also include uncompressed keys?

Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp:48]: Good call, done.

File: src/test/fuzz/descriptor_parse.cpp [hunk 2]
```diff
@@ -3,11 +3,111 @@
...
+
+            // If this is a "raw" key, generate a normal privkey. Otherwise generate
+            // an extended one.
+            if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i) || IdIsXOnlyPubKey(i) || IdIsConstPrivKey(i)) {
+                CKey privkey;
+                privkey.Set(key_data, key_data + 32, true);
+                if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i)) {
+                    CPubKey pubkey{privkey.GetPubKey()};
+                    if (IdIsUnCompPubKey(i)) assert(pubkey.Decompress());
+                    keys_str[i] = HexStr(pubkey);
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp:52]: Done.
```patch
diff --git a/src/test/fuzz/descriptor_parse.cpp b/src/test/fuzz/descriptor_parse.cpp
index 9a7b6ea308..529dadadc1 100644
--- a/src/test/fuzz/descriptor_parse.cpp
+++ b/src/test/fuzz/descriptor_parse.cpp
@@ -45,10 +45,9 @@ public:
             // an extended one.
             if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i) || IdIsXOnlyPubKey(i) || IdIsConstPrivKey(i)) {
                 CKey privkey;
-                privkey.Set(key_data, key_data + 32, true);
+                privkey.Set(key_data, key_data + 32, !IdIsUnCompPubKey(i));
                 if (IdIsCompPubKey(i) || IdIsUnCompPubKey(i)) {
                     CPubKey pubkey{privkey.GetPubKey()};
-                    if (IdIsUnCompPubKey(i)) assert(pubkey.Decompress());
                     keys_str[i] = HexStr(pubkey);
                 } else if (IdIsXOnlyPubKey(i)) {
                     const XOnlyPubKey pubkey{privkey.GetPubKey()};
```

**Comment by luke-jr** [comment]
Suggest putting "Fuzz" in title, and labelling.

**Comment by darosior** [comment]
Thanks for testing. Fixed the issue. I also introduced the bug you shared and it does make the target crash.

**Review by maflcko — APPROVED** [review]
nice, lgtm ACK 84dee4fe690e08a5adaad1c78530666da07075d8 🦄

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: nice, lgtm ACK 84dee4fe690e08a5adaad1c78530666da07075d8 🦄
ag2blJR9VoWl2NoAXwUZSDU54s5hPaFz7IJt+iLlaSMLYtYwbr4/XMw4CW9xm1A+BXAn9ZXtWQxrwFSRr/COBg==
```

</details>

File: src/test/fuzz/descriptor_parse.cpp [hunk 3]
```diff
@@ -3,28 +3,167 @@
...
+                } else if (IdIsXOnlyPubKey(i)) {
+                    const XOnlyPubKey pubkey{privkey.GetPubKey()};
+                    keys_str[i] = HexStr(pubkey);
+                } else {
+                    keys_str[i] = EncodeSecret(privkey);
+                }
+            } else {
+                CExtKey ext_privkey;
+                ext_privkey.SetSeed({(std::byte *)key_data, 32});
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp:60]: It's not avoiding them, it's hiding them behind `UCharCast`? :p 

It boils down to choosing between casting to `uint8_t` or to `std::byte`, and i went for the option with less casts.

File: src/test/fuzz/descriptor_parse.cpp [hunk 4]
```diff
@@ -3,28 +3,167 @@
...
+        if (idx.size() != 1) return {};
+        return idx[0];
+    }
+
+    //! Get a valid descriptor string from a descriptor string whose keys were mocked.
+    std::optional<std::string> GetDescriptor(std::string_view mocked_desc) const {
+        // The smallest fragment would be "pk(%00)"
+        if (mocked_desc.size() < 7) return {};
+
+        // The actual, valid, descriptor string to be returned.
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp]: Done, thanks. (And also in the function's doc comment.)

File: src/test/fuzz/descriptor_parse.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp:60]: Happy to do it then.

[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp]: Done.

(However i just checked and `UCharCast` does indeed use C-style casts)
https://github.com/bitcoin/bitcoin/blob/d23fda05842ba4539b225bbab01b94df0060f697/src/span.h#L272

File: src/test/fuzz/descriptor_parse.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp]: :facepalm:

**Comment by achow101** [comment]
ACK 131314b62e899f95d1863083d303b489b3212b16

File: src/test/fuzz/descriptor_parse.cpp [hunk 5]
```diff
@@ -3,28 +3,167 @@
...
+        }
+    }
+
+    //! Parse an id in the keys vectors from a 2-characters hex string.
+    std::optional<uint8_t> IdxFromHex(std::string_view hex_characters) const {
+        if (hex_characters.size() != 2) return {};
+        auto idx = ParseHex(hex_characters);
+        if (idx.size() != 1) return {};
+        return idx[0];
+    }
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/test/fuzz/descriptor_parse.cpp:77]: Maybe the target returning early on non-ASCII has basically the same effect?
